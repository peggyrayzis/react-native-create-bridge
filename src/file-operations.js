const path = require("path");
const fs = require("mz/fs");
const compareVersions = require("compare-versions");
const semver = require("../utils/semver");

const pkg = require(path.join(process.cwd(), "package.json"));

function getFileNames(dirPath) {
  return fs.readdir(dirPath).catch(e => console.error("[getFileNames] ", e));
}

function readFile(file, readDirPath) {
  return fs
    .readFile(path.join(readDirPath, file), "utf-8")
    .catch(e => console.error("[readFile] ", e));
}

function parseFile(fileData, { templateName, packageName, app, rnVersion }) {
  let kotlinPackage;
  let javaPackage;
  let version;

  try {
    version = semver(rnVersion);
  } catch (e) {
    version = "0.0.0";
  }

  // TODO: figure out a better way to handle one off breaking changes
  if (rnVersion && compareVersions(version, "0.47.2") < 0) {
    kotlinPackage = `
    override fun createJSModules(): List<Class<out JavaScriptModule>> {
        return emptyList()
    }

    `;

    javaPackage = `
    @Override
    public List<Class<? extends JavaScriptModule>> createJSModules() {
        return Collections.emptyList();
    }

    `;
  } else {
    kotlinPackage = "";
    javaPackage = "";
  }

  return fileData
    .replace(/{{template}}/g, templateName)
    .replace(/{{packageName}}/g, packageName)
    .replace(/{{app}}/g, app)
    .replace(/{{kotlinPackage}}/g, kotlinPackage)
    .replace(/{{javaPackage}}/g, javaPackage);
}

function readAndWriteFiles(files, paths, config) {
  const { readDirPath, writeDirPath } = paths;
  return Promise.all(
    files.map(file => {
      readFile(file, readDirPath).then(fileData => {
        const parsedFile = parseFile(fileData, config);
        const fileName = file.replace("Template", config.templateName);
        return fs.writeFile(path.join(writeDirPath, fileName), parsedFile);
      });
    })
  ).catch(e => console.error("[readAndWriteFiles] ", e));
}

module.exports = {
  pkg,
  getFileNames,
  readFile,
  parseFile,
  readAndWriteFiles
};
