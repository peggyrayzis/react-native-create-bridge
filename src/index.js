#! /usr/bin/env node

import inquirer from "inquirer";
import path from "path";
import isValid from "is-valid-path";
import mkdir from "mkdirp-promise";
import { success as successIcon, error as errorIcon } from "log-symbols";

import readAndWriteFiles, { pkg, getFileNames } from "./file-operations";

const templateNameRegex = /\w+/;
const promptConfig = [
  {
    type: "input",
    name: "templateName",
    message: "What is your bridge module called?",
    default: "ExampleBridge",
    validate: input => templateNameRegex.test(input)
  },
  {
    type: "checkbox",
    name: "bridgeType",
    message: "What type of bridge would you like to create?",
    default: ["Native Module", "Native UI Component"],
    choices: ["Native Module", "Native UI Component"]
  },
  {
    type: "checkbox",
    name: "environment",
    message: "What OS & languages would you like to support?",
    default: ["Android/Java", "iOS/Objective-C"],
    choices: ["Android/Java", "Android/Kotlin", "iOS/Swift", "iOS/Objective-C"]
  },
  {
    type: "input",
    name: "jsPath",
    message: "What directory should we deliver your JS files to?",
    default: ".",
    validate: input => isValid(input)
  }
];

const environmentMap = {
  "Android/Java": createJavaEnvironment,
  "Android/Kotlin": createKotlinEnvironment,
  "iOS/Swift": createSwiftEnvironment,
  "iOS/Objective-C": createObjCEnvironment
};

async function init() {
  try {
    const {
      environment,
      bridgeType,
      templateName,
      jsPath
    } = await inquirer.prompt(promptConfig);

    const templateFolder = bridgeType.length > 1
      ? "combined"
      : bridgeType[0] === "Native Module" ? "modules" : "ui-components";

    const promises = environment.map(env =>
      environmentMap[env](templateName, templateFolder)
    );

    promises.push(createJSEnvironment(templateName, templateFolder, jsPath));
    await Promise.all(promises);

    console.log(
      `${successIcon} Your bridge module was successfully created! ðŸŽ‰`
    );
  } catch (e) {
    console.log(
      `${errorIcon} Oh no! ðŸ’©  Something went wrong with creating your bridge module.\nPlease report any errors here: https://github.com/peggyrayzis/react-native-create-bridge/issues\n\nError: ${e}`
    );
  }
}

async function createJavaEnvironment(templateName, templateFolder) {
  const appPath = path.join(
    process.cwd(),
    "android",
    "app",
    "src",
    "main",
    "java",
    "com",
    pkg.name
  );

  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "android-java"
  );

  const writeDirPath = await mkdir(
    path.join(appPath, templateName.toLowerCase())
  );

  const paths = {
    readDirPath,
    writeDirPath: writeDirPath
      ? writeDirPath
      : path.join(appPath, templateName.toLowerCase())
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(
    files,
    paths,
    templateName,
    templateName.toLowerCase(),
    pkg.name.toLowerCase()
  );
}

async function createKotlinEnvironment(templateName, templateFolder) {
  const appPath = path.join(
    process.cwd(),
    "android",
    "app",
    "src",
    "main",
    "java",
    "com",
    pkg.name
  );

  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "android-kotlin"
  );

  const writeDirPath = await mkdir(
    path.join(appPath, templateName.toLowerCase())
  );

  const paths = {
    readDirPath,
    writeDirPath: writeDirPath
      ? writeDirPath
      : path.join(appPath, templateName.toLowerCase())
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(
    files,
    paths,
    templateName,
    templateName.toLowerCase(),
    pkg.name.toLowerCase()
  );
}

async function createSwiftEnvironment(templateName, templateFolder) {
  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "ios-swift"
  );

  const paths = {
    readDirPath,
    writeDirPath: path.join(process.cwd(), "ios")
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(files, paths, templateName);
}

async function createObjCEnvironment(templateName, templateFolder) {
  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "ios-objc"
  );

  const paths = {
    readDirPath,
    writeDirPath: path.join(process.cwd(), "ios")
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(files, paths, templateName);
}

async function createJSEnvironment(templateName, templateFolder, jsPath) {
  const readDirPath = path.join(
    __dirname,
    "..",
    "templates",
    templateFolder,
    "js"
  );

  await mkdir(jsPath);

  const paths = {
    readDirPath,
    writeDirPath: jsPath
  };

  const files = await getFileNames(readDirPath);

  return readAndWriteFiles(files, paths, templateName);
}

init();
