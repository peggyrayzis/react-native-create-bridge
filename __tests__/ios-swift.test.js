import path from 'path';
import { readFile, parseFile } from '../src/file-operations';

describe('iOS/Swift: UI Components', () => {
  const templateName = 'TestModule';
  const readDirPath = path.join(
    __dirname,
    '..',
    'templates',
    'ui-components',
    'ios-swift',
  );

  it('creates a Template-Bridging-Header.h', async () => {
    const fileData = await readFile('Template-Bridging-Header.h', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a Template.m', async () => {
    const fileData = await readFile('Template.m', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplateManager.swift', async () => {
    const fileData = await readFile('TemplateManager.swift', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplateView.swift', async () => {
    const fileData = await readFile('TemplateView.swift', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });
});

describe('iOS/Swift: Native Modules', () => {
  const templateName = 'TestModule';
  const readDirPath = path.join(
    __dirname,
    '..',
    'templates',
    'modules',
    'ios-swift',
  );

  it('creates a Template-Bridging-Header.h', async () => {
    const fileData = await readFile('Template-Bridging-Header.h', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a Template.m', async () => {
    const fileData = await readFile('Template.m', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a Template.swift', async () => {
    const fileData = await readFile('Template.swift', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });
});

describe('iOS/Swift: Combined', () => {
  const templateName = 'TestModule';
  const readDirPath = path.join(
    __dirname,
    '..',
    'templates',
    'combined',
    'ios-swift',
  );

  it('creates a Template-Bridging-Header.h', async () => {
    const fileData = await readFile('Template-Bridging-Header.h', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a Template.m', async () => {
    const fileData = await readFile('Template.m', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });

  it('creates a TemplateManager.swift', async () => {
    const fileData = await readFile('TemplateManager.swift', readDirPath);
    const parsedFile = parseFile(fileData, templateName);
    expect(parsedFile).toMatchSnapshot();
  });
});
