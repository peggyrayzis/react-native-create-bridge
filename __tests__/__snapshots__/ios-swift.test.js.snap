exports[`iOS/Swift: Combined creates a Template.m 1`] = `
"//  Created by react-native-create-bridge

// import RCTViewManager
#if __has_include(<React/RCTViewManager.h>)
#import <React/RCTViewManager.h>
#elif __has_include(“RCTViewManager.h”)
#import “RCTViewManager.h”
#else
#import “React/RCTViewManager.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventDispatcher
#if __has_include(<React/RCTEventDispatcher.h>)
#import <React/RCTEventDispatcher.h>
#elif __has_include(“RCTEventDispatcher.h”)
#import “RCTEventDispatcher.h”
#else
#import “React/RCTEventDispatcher.h” // Required when used as a Pod in a Swift project
#endif

// Export a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html#exporting-swift
@interface RCT_EXTERN_MODULE(TestModule, RCTViewManager)

// Map native properties to React Component props
// https://facebook.github.io/react-native/docs/native-components-ios.html#properties
RCT_EXPORT_VIEW_PROPERTY(\"exampleProp\", NSString)

// Export methods to a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html#exporting-swift
RCT_EXTERN_METHOD(exampleMethod)

@end
"
`;

exports[`iOS/Swift: Combined creates a Template-Bridging-Header.h 1`] = `
"//  Created by react-native-create-bridge
// TestModule-Bridging-Header.h

// import RCTBridgeModule
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import “RCTBridgeModule.h”
#else
#import “React/RCTBridgeModule.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventEmitter
#if __has_include(<React/RCTViewManager.h>)
#import <React/RCTViewManager.h>
#elif __has_include(“RCTViewManager.h”)
#import “RCTViewManager.h”
#else
#import “React/RCTViewManager.h” // Required when used as a Pod in a Swift project
#endif"
`;

exports[`iOS/Swift: Combined creates a TemplateManager.swift 1`] = `
"//  Created by react-native-create-bridge

import Foundation

@objc(TestModule)
class TestModuleManager : RCTViewManager {
  // Export constants to use in your native module
  override func constantsToExport() -> [String : Any]! {
    return [\"EXAMPLE_CONSTANT\": \"example\"]
  }

  // Return the native view that represents your React component
  override func view() -> UIView! {
    return UIView()
  }

  // Implement methods that you want to export to the native module
  @objc func exampleMethod() {
    // The bridge eventDispatcher is used to send events from native to JS env
    // No documentation yet on DeviceEventEmitter: https://github.com/facebook/react-native/issues/2819
    self.bridge.eventDispatcher().sendAppEvent(withName: \"EXAMPLE_EVENT\", body: nil)
  }
}
"
`;

exports[`iOS/Swift: Native Modules creates a Template.m 1`] = `
"//  Created by react-native-create-bridge

// import RCTBridgeModule
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import “RCTBridgeModule.h”
#else
#import “React/RCTBridgeModule.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventEmitter
#if __has_include(<React/RCTEventEmitter.h>)
#import <React/RCTEventEmitter.h>
#elif __has_include(“RCTEventEmitter.h”)
#import “RCTEventEmitter.h”
#else
#import “React/RCTEventEmitter.h” // Required when used as a Pod in a Swift project
#endif

// Export a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html#exporting-swift
@interface RCT_EXTERN_MODULE(TestModule, NSObject)

// Export methods to a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html#exporting-swift
RCT_EXTERN_METHOD(exampleMethod)

@end
"
`;

exports[`iOS/Swift: Native Modules creates a Template.swift 1`] = `
"//  Created by react-native-create-bridge

import Foundation

@objc(TestModule)
class TestModule : NSObject {
  // Export constants to use in your native module
  override func constantsToExport() -> [String : Any]! {
    return [\"EXAMPLE_CONSTANT\": \"example\"]
  }

  // Implement methods that you want to export to the native module
  @objc func exampleMethod() {
    // write method here
  }
}
"
`;

exports[`iOS/Swift: Native Modules creates a Template-Bridging-Header.h 1`] = `
"//  Created by react-native-create-bridge
// TestModule-Bridging-Header.h

// import RCTBridgeModule
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import “RCTBridgeModule.h”
#else
#import “React/RCTBridgeModule.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventEmitter
#if __has_include(<React/RCTEventEmitter.h>)
#import <React/RCTEventEmitter.h>
#elif __has_include(“RCTEventEmitter.h”)
#import “RCTEventEmitter.h”
#else
#import “React/RCTEventEmitter.h” // Required when used as a Pod in a Swift project
#endif"
`;

exports[`iOS/Swift: UI Components creates a Template.m 1`] = `
"//  Created by react-native-create-bridge

// import RCTViewManager
#if __has_include(<React/RCTViewManager.h>)
#import <React/RCTViewManager.h>
#elif __has_include(“RCTViewManager.h”)
#import “RCTViewManager.h”
#else
#import “React/RCTViewManager.h” // Required when used as a Pod in a Swift project
#endif

// Export a native module
// https://facebook.github.io/react-native/docs/native-modules-ios.html#exporting-swift
@interface RCT_EXTERN_MODULE(TestModule, RCTViewManager)

// Map native properties to React Component props
// https://facebook.github.io/react-native/docs/native-components-ios.html#properties
RCT_EXPORT_VIEW_PROPERTY(\"exampleProp\", NSString)

@end
"
`;

exports[`iOS/Swift: UI Components creates a Template-Bridging-Header.h 1`] = `
"//  Created by react-native-create-bridge
// TestModule-Bridging-Header.h

// import RCTBridgeModule
#if __has_include(<React/RCTBridgeModule.h>)
#import <React/RCTBridgeModule.h>
#elif __has_include(“RCTBridgeModule.h”)
#import “RCTBridgeModule.h”
#else
#import “React/RCTBridgeModule.h” // Required when used as a Pod in a Swift project
#endif

// import RCTEventEmitter
#if __has_include(<React/RCTViewManager.h>)
#import <React/RCTViewManager.h>
#elif __has_include(“RCTViewManager.h”)
#import “RCTViewManager.h”
#else
#import “React/RCTViewManager.h” // Required when used as a Pod in a Swift project
#endif"
`;

exports[`iOS/Swift: UI Components creates a TemplateManager.swift 1`] = `
"//  Created by react-native-create-bridge

import Foundation
import UIKit

@objc(TestModule)
class TestModuleManager : RCTViewManager {
  // Return the native view that represents your React component
  override func view() -> UIView! {
    return TestModuleView()
  }

  // Export constants to use in your native module
  override func constantsToExport() -> [String : Any]! {
    return [\"EXAMPLE_CONSTANT\": \"example\"]
  }

}
"
`;

exports[`iOS/Swift: UI Components creates a TemplateView.swift 1`] = `
"//  Created by react-native-create-bridge

import Foundation

@objc(TestModuleView)
class TestModuleView : UIView {
  
  override init(frame: CGRect) {
    super.init(frame: frame);
    self.frame = frame;
  }

  required init?(coder aDecoder: NSCoder) {
      fatalError(\"init(coder:) has not been implemented\")
  }

  func setExampleProp(exampleProp: NSString) {
    self.exampleProp = exampleProp
  }

}"
`;
